clang/lib/Driver/ToolChains/Cygwin.cpp

```diff
@@ -107,3 +110,300 @@ void Cygwin::AddClangSystemIncludeArgs(const ArgList &DriverArgs,
+  if (!Args.hasArg(options::OPT_shared) && !Args.hasArg(options::OPT_mdll)) {
+    if (ToolChain.getTriple().isArch32Bit())
+      CmdArgs.push_back("--large-address-aware");
+    CmdArgs.push_back("--tsaware");
+  }
+
+  CmdArgs.push_back("-o");
+  const char *OutputFile = Output.getFilename();
+  // GCC implicitly adds an .exe extension if it is given an output file name
+  // that lacks an extension.
+  // GCC used to do this only when the compiler itself runs on windows, but
+  // since GCC 8 it does the same when cross compiling as well.
+  if (!llvm::sys::path::has_extension(OutputFile)) {
+    CmdArgs.push_back(Args.MakeArgString(Twine(OutputFile) + ".exe"));
+    OutputFile = CmdArgs.back();
+  } else
+    CmdArgs.push_back(OutputFile);
+
+  if (!Args.hasArg(options::OPT_nostdlib, options::OPT_nostartfiles,
+                   options::OPT_r)) {
+    const bool IsShared = Args.hasArg(options::OPT_shared);
+    if (Args.hasArg(options::OPT_mdll) || IsShared) {
+      CmdArgs.push_back("-e");
+      CmdArgs.push_back(ToolChain.getArch() == llvm::Triple::x86
+                            ? "__cygwin_dll_entry@12"
+                            : "_cygwin_dll_entry");
+      CmdArgs.push_back("--enable-auto-image-base");
+    }
+
+    if (!Args.hasArg(options::OPT_mdll) && !IsShared)
+      CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath("crt0.o")));
+    if (ToolChain.GetRuntimeLibType(Args) == ToolChain::RLT_CompilerRT) {
+      std::string crtbegin =
+          ToolChain.getCompilerRT(Args, "crtbegin", ToolChain::FT_Object);
+      if (ToolChain.getVFS().exists(crtbegin)) {
+        std::string P;
+        P = crtbegin;
+        CmdArgs.push_back(Args.MakeArgString(P));
+      }
+    }
+    if (IsShared)
+      CmdArgs.push_back(
+          Args.MakeArgString(ToolChain.GetFilePath("crtbeginS.o")));
+    else
+      CmdArgs.push_back(
+          Args.MakeArgString(ToolChain.GetFilePath("crtbegin.o")));
+
+    // Add crtfastmath.o if available and fast math is enabled.
+    ToolChain.addFastMathRuntimeIfAvailable(Args, CmdArgs);
+  }
+
+  Args.addAllArgs(CmdArgs, {options::OPT_L, options::OPT_u});
+
+  ToolChain.AddFilePathLibArgs(Args, CmdArgs);
+
+  if (D.isUsingLTO())
+    tools::addLTOOptions(ToolChain, Args, CmdArgs, Output, Inputs,
+                         D.getLTOMode() == LTOK_Thin);
+
+  if (Args.hasArg(options::OPT_Z_Xlinker__no_demangle))
+    CmdArgs.push_back("--no-demangle");
+
+  bool NeedsSanitizerDeps =
+      tools::addSanitizerRuntimes(ToolChain, Args, CmdArgs);
+  bool NeedsXRayDeps = tools::addXRayRuntime(ToolChain, Args, CmdArgs);
+  tools::addLinkerCompressDebugSectionsOption(ToolChain, Args, CmdArgs);
+  tools::AddLinkerInputs(ToolChain, Inputs, Args, CmdArgs, JA);
+
+  bool saw_high_entropy_va = false;
+  bool saw_nxcompat = false;
+  for (const char *Arg : CmdArgs) {
+    if (StringRef(Arg) == "-high-entropy-va" ||
+        StringRef(Arg) == "--high-entropy-va" ||
```

Comment:
I wonder if there's a more clever way to express this, without having to repeat every constant twice, for checking for one or two leading dashes? And the `StringRef(Arg)` wrapping maybe could be stored in a variable instead of remade in each comparison?

Reply:
thinking about this more, these are last-wins options, so couldn't we just add our intended defaults before the AddLinkerInputs and be done with it?  Or am I forgetting the reason why that wouldn't work?

Reply:
I think that should work fine as well.